File: android_quickbrick_app/src/main/java/com/applicaster/ui/activities/MainActivity.java
package com.applicaster.ui.activities;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.view.KeyEvent;
import android.view.OrientationEventListener;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.annotation.RawRes;
import androidx.core.view.ViewCompat;

import com.applicaster.ui.interfaces.HostActivityBase;
import com.applicaster.ui.interfaces.IUILayerManager;
import com.applicaster.ui.loaders.PreloadStateManager;
import com.applicaster.ui.loaders.PreloadStateManager.PreloadStep;
import com.applicaster.ui.quickbrick.QuickBrickManager;
import com.applicaster.ui.utils.OrientationUtils;
import com.applicaster.util.APLogger;
import com.applicaster.util.AppData;
import com.applicaster.util.OSUtil;
import com.applicaster.util.UrlSchemeUtil;
import com.applicaster.util.ui.APUIUtils;
import com.applicaster.util.ui.ApplicationPreloader;
import com.applicaster.util.ui.PreloaderListener;
import com.applicaster.zapp.quickbrick.loader.DataLoader;

import java.util.Stack;

import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class MainActivity extends HostActivityBase {

    private static final String TAG = MainActivity.class.getSimpleName();

    private static final String INTRO_LAYOUT = "intro";
    private static final String INTRO_VIDEO_RAW_RESOURCE_DEFAULT = "intro";
    private static final String INTRO_VIDEO_RAW_RESOURCE_TABLET = "intro_tablet";
    private static final String APPLICATION_PRELOADER_LAYOUT_ID = "preloader_view";

    private final Stack<Integer> orientationStack = new Stack<>();

    private PreloadStateManager preloadStateManager;
    private IUILayerManager uiLayer;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        preloadStateManager = new PreloadStateManager();
        setAppOrientation();
        setSplashAndApplicationPreloaderView();
        playVideoIntroIfPresent();
        // todo: show debug setup dialog here
        loadData();
        // todo: check if it can run in BG
        new Handler().post(this::initializeUILayer);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        if(null != uiLayer && uiLayer.isInitialized()) {
            Uri uri = UrlSchemeUtil.getUrlSchemeData(intent);
            if (null != uri) {
                uiLayer.handleURL(uri.toString());
            }
        }
    }

    private void loadData() {
        // todo: move everything into single observable and remove preloadStateManager
        DataLoader.initialize(this, getIntent())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()).subscribe(
                        () -> preloadStepComplete(PreloadStateManager.PreloadStep.LOAD_DATA),
                        throwable -> APLogger.error(TAG, "QuickBrickActivity load data failed", throwable));
    }

    private void initOrientationListener() {
        new OrientationEventListener(this) {

            private int lastKnownRotation = 0;

            {
                if (canDetectOrientation()) {
                    enable();
                }
            }

            /**
             * This is either Surface.Rotation_0, _90, _180, _270, or -1 (invalid).
             */
            @Override
            public void onOrientationChanged(int orientation) {
                int normalisedOrientation = OrientationUtils.INSTANCE.normaliseOrientation(orientation);
                if (!OrientationUtils.INSTANCE.supportOrientation(normalisedOrientation, orientationStack.peek())) {
                    return;
                }
                if (lastKnownRotation == normalisedOrientation) {
                    return;
                }
                int from = OrientationUtils.INSTANCE.jsOrientationMapper(lastKnownRotation);
                lastKnownRotation = normalisedOrientation;
                int to = OrientationUtils.INSTANCE.jsOrientationMapper(normalisedOrientation);
                if (uiLayer != null && uiLayer.isInitialized()) {
                    uiLayer.orientationChange(from, to);
                }
            }
        };
    }

    // region activity life cycle events

    /**
     * Handle onKeyDown in debug builds only
     */
    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        boolean shouldOverrideKeyDownEvent = false;
        if (uiLayer != null && uiLayer.isInitialized()) {
            shouldOverrideKeyDownEvent = uiLayer.onKeyDown(keyCode, event);
        }
        return shouldOverrideKeyDownEvent || super.onKeyDown(keyCode, event);
    }

    /**
     * Let RN handle back button press, only after initialize is complete
     */
    @Override
    public void onBackPressed() {
        if (uiLayer != null && uiLayer.isInitialized()) {
            uiLayer.onBackPressed();
        } else {
            super.onBackPressed();
        }
    }

    /**
     * RN: Resume responding to touch events and re-register various event listeners.
     */
    @Override
    protected void onResume() {
        super.onResume();
        if (uiLayer != null && uiLayer.isInitialized()) uiLayer.onResume();
    }

    /**
     * AppData: persist state in Preferences.
     * RN: un-register various event listeners & some cleanup.
     */
    @Override
    protected void onPause() {
        super.onPause();
        AppData.persistAppData(this);
        if (uiLayer != null && uiLayer.isInitialized())uiLayer.onPause();
    }

    /**
     * RN: stop RN's instance manager and detach RN's application from RN's root view.
     */
    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (uiLayer != null) uiLayer.onDestroy();
    }
    //endregion

    private void setAppOrientation() {
        APUIUtils.setOrientation(this);
        orientationStack.add(this.getRequestedOrientation());
    }

    @Override
    public void setAppOrientation(int orientation) {
        int nativeOrientation = OrientationUtils.INSTANCE.nativeOrientationMapper(orientation);
        this.setRequestedOrientation(nativeOrientation);
        orientationStack.add(this.getRequestedOrientation());
    }

    @Override
    public void releaseOrientation() {
        if (orientationStack.size() > 1) {
            orientationStack.pop();
            this.setRequestedOrientation(orientationStack.peek());
        }
    }

    /**
     * intro.xml creates:
     * - intro_splash_layout containing @drawable/splash_logo
     * - ApplicationPreloader layout, which both creates a placeholder for
     * either image, webview or video, AND handles the dismiss behavior
     * (Yes! a linearLayout that makes decisions, brilliant)
     */
    private void setSplashAndApplicationPreloaderView() {
        setContentView(OSUtil.getLayoutResourceIdentifier(INTRO_LAYOUT));
        View root = getWindow().getDecorView().getRootView();
        ViewCompat.setOnApplyWindowInsetsListener(root, (v, insets) -> insets.consumeSystemWindowInsets());
    }

    /**
     * Update preload step complete and check if all steps are complete.
     * If true - set content view to RN's root view, replacing splash screen layout.
     * If preload was already complete before this call, return -
     * this is a safeguard against rogue events that are accidentally sent more than once.
     *
     * @param step Preload step like video play complete, RN initialize complete, etc.
     */
    private synchronized void preloadStepComplete(PreloadStep step) {
        if(isFinishing())
            return;

        if (preloadStateManager.isPreloadComplete())
            return; // preload was already complete, don't do anything

        preloadStateManager.setStepComplete(step);

        if (preloadStateManager.isPreloadComplete()) {
            runOnUiThread(() -> {
                uiLayer.onResume(); // This placement is intentional! can only be run inside a UI thread.
                setContentView(uiLayer.getRootView()); // simplistic approach, replace whole intro layout with RN layout
                Uri uri = UrlSchemeUtil.getUrlSchemeData(getIntent());
                if(null != uri) {
                    uiLayer.handleURL(uri.toString());
                }
            });
        }
    }

    /**
     * Check if video intro resource exists.
     * - Exists? Play video and connect onIntroVideoFinished to PreloadStep.VIDEO_INTRO
     * - Does not exist? Set PreloadStep.VIDEO_INTRO as complete
     */
    private void playVideoIntroIfPresent() {
        if (getVideoIntroResource() == 0) {
            preloadStepComplete(PreloadStep.VIDEO_INTRO);
            return;
        }

        ApplicationPreloader applicationPreloaderView = findViewById(OSUtil.getResourceId(APPLICATION_PRELOADER_LAYOUT_ID));
        applicationPreloaderView.setListener(new PreloaderViewListener(
                () -> preloadStepComplete(PreloadStep.VIDEO_INTRO))
        );

        Uri videoIntroUri = buildVideoIntroUri(getVideoIntroResource());
        applicationPreloaderView.showIntro(videoIntroUri, this);
    }

    /**
     * VideoView requires a Uri scheme for the location of the video resource.
     *
     * @param videoIntroResourceID Resource in the res/raw folder.
     * @return Uri representation of the packaged video file location.
     */
    private Uri buildVideoIntroUri(@RawRes int videoIntroResourceID) {
        return Uri.parse("android.resource://" + getPackageName() + "/" + videoIntroResourceID);
    }

    /**
     * Creates the QuickBrick manager that holds the RN view,
     * and handles the initialization process.
     * The Manager interacts with this activity using this interface:
     * {@link IUILayerManager.StatusListener}
     */
    private void initializeUILayer() {
        if(isFinishing()) {
            return;
        }
        uiLayer = new QuickBrickManager(this);
        uiLayer.setEventsListener(new IUILayerManager.StatusListener() {
            @Override
            public void onReady() {
                initOrientationListener();
                preloadStepComplete(PreloadStep.UI_LAYER);
            }

            @Override
            public void onError(@Nullable Exception e) {
                Log.e(TAG, "QuickBrickManager error", e);

                Handler handler = new Handler();
                handler.postDelayed(() -> {
                    finish(); // Not very nice but we prefer failing hard and fast in this case
                }, 1000);
            }
        });
        uiLayer.start();
    }

    /**
     * Check for device type, and return resource for tablet video, if available.
     * Default: return default video resource (phone/portrait video)
     *
     * @return resource id of intro video
     */
    @RawRes
    private int getVideoIntroResource() {
        int resourceId = OSUtil.getRawResourceIdentifier(INTRO_VIDEO_RAW_RESOURCE_DEFAULT);
        if (OSUtil.isTablet()) {
            int tabletResourceId = OSUtil.getRawResourceIdentifier(INTRO_VIDEO_RAW_RESOURCE_TABLET);
            if (tabletResourceId > 0) {
                resourceId = tabletResourceId;
            }
        }
        return resourceId;
    }

    static class PreloaderViewListener implements PreloaderListener {

        private final Runnable callback;

        public PreloaderViewListener(Runnable callback) {
            this.callback = callback;
        }

        /**
         * Will be called several times during pre-load process,
         * in no logical/useful manner. Just ignore everything.
         */
        @Override
        public void onPreloaderStart() {
        }

        /**
         * Handles other "onComplete" events in the {@link ApplicationPreloader},
         * like dismissed interstitial / ad / webview
         */
        @Override
        public void onPreloaderFinish() {
        }

        /**
         * NEVER called
         * @param e Exception from {@link ApplicationPreloader}
         */
        @Override
        public void handlePreloaderException(Exception e) {
        }

        /**
         * Run callback when there is nothing left to do from the video playing perspective, either:
         * 1. There's an error in the player (error is swallowed, screw you)
         * 2. Video completed playing (VideoView visibility is already set as "gone")
         */
        @Override
        public void onIntroVideoFinished() {
            callback.run();
        }
    }
}

File: android_quickbrick_app/src/main/java/com/applicaster/ui/activities/MainActivity.kt
package com.applicaster.ui.activities

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.text.TextUtils
import android.view.KeyEvent
import android.view.OrientationEventListener
import android.widget.Toast
import androidx.annotation.RawRes
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.OnLifecycleEvent
import com.applicaster.analytics.AnalyticsAgentUtil
import com.applicaster.plugin_manager.PluginManager
import com.applicaster.plugin_manager.cmp.ConsentManager
import com.applicaster.ui.R
import com.applicaster.ui.interfaces.HostActivityBase
import com.applicaster.ui.interfaces.IUILayerManager
import com.applicaster.ui.loaders.MainActivityPreloadSequence
import com.applicaster.ui.quickbrick.QuickBrickManager
import com.applicaster.ui.utils.AppHookExecutor
import com.applicaster.ui.utils.ConsentHookExecutor
import com.applicaster.ui.utils.OrientationUtils.jsOrientationMapper
import com.applicaster.ui.utils.OrientationUtils.nativeOrientationMapper
import com.applicaster.ui.utils.OrientationUtils.normaliseOrientation
import com.applicaster.ui.utils.OrientationUtils.supportOrientation
import com.applicaster.ui.views.ApplicationPreloaderView
import com.applicaster.util.*
import com.applicaster.util.ui.APUIUtils
import com.applicaster.util.ui.PreloaderListener
import io.reactivex.Completable
import io.reactivex.CompletableEmitter
import java.util.*

class MainActivity : HostActivityBase() {

    private val orientationStack = Stack<Int>()
    private var uiLayer: IUILayerManager? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        APLogger.debug(TAG, "Activity onCreate called")
        super.onCreate(savedInstanceState)
        if (routeOrUpdateIntent(intent)) {
            // Abort launch if the intent was supposed to be handled in external application.
            // Can't use finishAndRemoveTask() since external app we launch is in the same stack, and will be terminated too
            // So our app still will be visible in the recent apps history.
            finish()
            return
        }
        setAppOrientation()
        setContentView(R.layout.intro)
        // todo: show debug setup dialog here
        MainActivityPreloadSequence.configure(this).run()
    }

    private fun routeOrUpdateIntent(intent: Intent): Boolean {
        // Take care of url extra delivered from Firebase Console Firebase Push when app was not active
        // if url schema belongs to our app, we will update the intent,
        // otherwise we will try to route it as it was supposed to be done.
        val data = intent.data
        APLogger.info(TAG, "Intent received with data: " + (data?.toString() ?: "null"))
        if (!intent.hasExtra(INTENT_EXTRA_URL)) {
            return false
        }
        val url = intent.getStringExtra(INTENT_EXTRA_URL)
        if (TextUtils.isEmpty(url)) {
            return false
        }
        APLogger.info(TAG, "Received Url extra: $url")
        val uri = Uri.parse(url)
        return if (UrlSchemeUtil.isUrlScheme(url)) {
            // update intent in-place, and let caller proceed
            intent.data = uri
            setIntent(Intent(Intent.ACTION_VIEW, uri))
            APLogger.info(TAG, "Intent data was replaced with url extra: $url")
            false
        } else {
            val externalIntent = Intent(Intent.ACTION_VIEW, uri)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            startActivity(externalIntent)
            APLogger.info(TAG, "Intent was routed to an external application")
            true
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent) // save new intent in any case
        if (routeOrUpdateIntent(intent)) {
            // This was probably intent from Firebase Console Firebase Push when app was not active.
            // User has first launched the app, and then clicked the notification
            return
        }
        // Update current intent in any case,
        // since QB may still be launching, and will only check it later.
        if (true == uiLayer?.isReady) {
            // get intent back since routeOrUpdateIntent can update it
            val uri = UrlSchemeUtil.getUrlSchemeData(getIntent())
            if (null != uri) {
                uiLayer!!.handleURL(uri.toString())
            }
        }
    }

    private fun initOrientationListener() {
        object : OrientationEventListener(this), LifecycleObserver {
            private var lastKnownRotation = 0

            /**
             * This is either Surface.Rotation_0, _90, _180, _270, or -1 (invalid).
             */
            override fun onOrientationChanged(orientation: Int) {
                val normalisedOrientation = normaliseOrientation(orientation)
                if (!supportOrientation(normalisedOrientation, orientationStack.peek())) {
                    return
                }
                if (lastKnownRotation == normalisedOrientation) {
                    return
                }
                val from = jsOrientationMapper(lastKnownRotation)
                lastKnownRotation = normalisedOrientation
                val to = jsOrientationMapper(normalisedOrientation)
                if (uiLayer != null && uiLayer!!.isReady) {
                    APLogger.info(TAG, "Reporting onOrientationChanged event to UI: $orientation")
                    uiLayer!!.orientationChange(from, to)
                }
            }

            @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
            fun onActivityDestroy() {
                disable()
            }

            init {
                if (canDetectOrientation()) {
                    enable()
                }
                lifecycle.addObserver(this)
            }
        }
    }
    // region activity life cycle events
    /**
     * Handle onKeyDown in debug builds only
     */
    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        var shouldOverrideKeyDownEvent = false
        if (true == uiLayer?.isReady) {
            shouldOverrideKeyDownEvent = uiLayer!!.onKeyDown(keyCode, event)
        }
        return shouldOverrideKeyDownEvent || super.onKeyDown(keyCode, event)
    }

    /**
     * Let RN handle back button press, only after initialize is complete
     */
    override fun onBackPressed() {
        if (true == uiLayer?.isReady) {
            uiLayer!!.onBackPressed()
        } else {
            super.onBackPressed()
        }
    }

    /**
     * RN: Resume responding to touch events and re-register various event listeners.
     */
    override fun onResume() {
        super.onResume()
        uiLayer?.onResume()
    }

    /**
     * AppData: persist state in Preferences.
     * RN: un-register various event listeners & some cleanup.
     */
    override fun onPause() {
        super.onPause()
        AppData.persistAppData(this)
        uiLayer?.onPause()
    }

    /**
     * RN: stop RN's instance manager and detach RN's application from RN's root view.
     */
    override fun onDestroy() {
        super.onDestroy()
        uiLayer?.onDestroy()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        uiLayer?.onActivityResult(requestCode, resultCode, data)
    }

    //endregion
    private fun setAppOrientation() {
        APUIUtils.setOrientation(this)
        orientationStack.add(this.requestedOrientation)
        APLogger.info(TAG, "Setting requested orientation: $requestedOrientation, stack depth ${orientationStack.size}")
    }

    override fun setAppOrientation(orientation: Int) {
        val nativeOrientation = nativeOrientationMapper(orientation)
        this.requestedOrientation = nativeOrientation
        orientationStack.add(this.requestedOrientation)
        APLogger.info(TAG, "Orientation change requested: $nativeOrientation, stack depth ${orientationStack.size}")
    }

    override fun releaseOrientation() {
        if (orientationStack.size > 1) {
            orientationStack.pop()
            this.requestedOrientation = orientationStack.peek()
            APLogger.info(TAG, "Orientation change released, stack depth ${orientationStack.size}")
        } else {
            // first one is startup one, should not be popped
            APLogger.warn(TAG, "Orientation change released, but stack is empty")
        }
    }

    internal fun executeHooks(isAppReady: Boolean) : Completable {
        val hookPluginList = PluginManager.getInstance().hookPluginList
        return when {
            hookPluginList == null || hookPluginList.isEmpty() -> Completable.complete()
            else -> Completable.create {
                AppHookExecutor(this,
                        hookPluginList,
                        it,
                        isAppReady)
            }
        }
    }

    /**
     * Check if video intro resource exists.
     * - Exists? Play video and connect onIntroVideoFinished to PreloadStep.VIDEO_INTRO
     * - Does not exist? Set PreloadStep.VIDEO_INTRO as complete
     */
    internal fun playVideoIntroIfPresent(): Completable {
        val introResource = videoIntroResource
        if (introResource == 0) {
            return Completable.complete()
        }
        return createUIThreadCompletable { completableEmitter ->
            findViewById<ApplicationPreloaderView>(R.id.preloader_view).let {
                it.setListener(PreloaderViewListener(completableEmitter))
                val videoIntroUri = buildVideoIntroUri(introResource)
                it.showIntro(videoIntroUri, this@MainActivity)
            }
        }
    }

    /**
     * VideoView requires a Uri scheme for the location of the video resource.
     *
     * @param videoIntroResourceID Resource in the res/raw folder.
     * @return Uri representation of the packaged video file location.
     */
    private fun buildVideoIntroUri(@RawRes videoIntroResourceID: Int): Uri {
        return Uri.parse("android.resource://$packageName/$videoIntroResourceID")
    }

    /**
     * Creates the QuickBrick manager that holds the RN view,
     * and handles the initialization process.
     * The Manager interacts with this activity using this interface:
     * [IUILayerManager.StatusListener]
     */
    internal fun initializeUILayer(): Completable {
        if (isFinishing) {
            return Completable.complete()
        }
        return createUIThreadCompletable { completableEmitter ->
            APLogger.debug(TAG, "Initializing UI layer...")
            uiLayer = QuickBrickManager(this@MainActivity)
            uiLayer!!.setEventsListener(object : IUILayerManager.StatusListener {
                override fun onReady() = completableEmitter.onComplete()

                override fun onError(e: Exception?) {
                    APLogger.error(TAG, "QuickBrickManager error: ${e ?: " (no exception)"}", e)
                    val handler = Handler(Looper.getMainLooper())
                    handler.post {
                        // post on UI thread, some devices has issues with toasts on worker ones
                        Toast.makeText(
                                AppContext.get(), // use app context so toast will survive past activity finish
                                "QuickBrickManager critical error: $e. The Application will now close.",
                                Toast.LENGTH_LONG).show()
                    }
                    handler.postDelayed({
                        finish() // Not very nice but we prefer failing hard and fast in this case
                    }, 1000)
                }
            })
            uiLayer!!.start()
        }
    }

    internal fun showUI(): Completable {
        if (isFinishing) {
            return Completable.complete()
        }
        return createUIThreadCompletable { completableEmitter ->
            APLogger.debug(TAG, "UI ready...")
            initOrientationListener()
            AnalyticsAgentUtil.logEvent(AnalyticsAgentUtil.APPLICATION_STARTED)
            setContentView(uiLayer!!.rootView) // simplistic approach, replace whole intro layout with RN layout
            completableEmitter.onComplete()
        }
    }

    // Not really needed right now, since we run callbacks on UI thread in PreloadStateManager.
    private fun createUIThreadCompletable(action: (CompletableEmitter) -> Unit): Completable {
        return Completable.create {
            runOnUiThread { action(it) }
        }
    }

    fun executeConsentHooks(): Completable {
        val consentPlugins = ConsentManager.getConsentPlugins()
        return when {
            consentPlugins.isEmpty() -> Completable.complete()
            else -> Completable.create {
                ConsentHookExecutor(this,
                        consentPlugins,
                        it)
            }
        }
    }

    /**
     * Check return resource for video, if available.
     * Default: return default video resource (phone/portrait video)
     *
     * @return resource id of intro video
     */
    @get:RawRes
    private val videoIntroResource: Int
        get() = OSUtil.getRawResourceIdentifier(INTRO_VIDEO_RAW_RESOURCE_DEFAULT)

    internal class PreloaderViewListener(private val callback: CompletableEmitter) : PreloaderListener {

        /**
         * Will be called several times during pre-load process,
         * in no logical/useful manner. Just ignore everything.
         */
        override fun onPreloaderStart() {}

        /**
         * Handles other "onComplete" events in the [ApplicationPreloaderView],
         * like dismissed interstitial / ad / webview
         */
        override fun onPreloaderFinish() {}

        /**
         * NEVER called
         * @param e Exception from [ApplicationPreloaderView]
         */
        override fun handlePreloaderException(e: Exception) {}

        /**
         * Run callback when there is nothing left to do from the video playing perspective, either:
         * 1. There's an error in the player (error is swallowed, screw you)
         * 2. Video completed playing (VideoView visibility is already set as "gone")
         */
        override fun onIntroVideoFinished() = callback.onComplete()
    }

    companion object {
        private const val TAG = "MainActivity"
        private const val INTRO_VIDEO_RAW_RESOURCE_DEFAULT = "intro"

        // intents from Firebase Push messages handled in the background will deliver url in the extras
        private const val INTENT_EXTRA_URL = "url"
    }
}